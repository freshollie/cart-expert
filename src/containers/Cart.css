.cart-details {
  width: 100%;
}

.cart-details > tbody > tr > td:last-child {
  text-align: right;
}

.cart-container {
  font-size: 20px;
  width: 100%;
  overflow-y: scroll;
}

.break {
  border-bottom: dashed;
  border-width: 2px;
}
