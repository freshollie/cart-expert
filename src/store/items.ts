import {
  createSelector,
  createSlice,
} from "@reduxjs/toolkit";
import { RootState } from ".";
import { Item } from "../types";

const ITEMS: readonly Item[] = [
  {
    id: "34583",
    name: "Beans",
    price: {
      amount: 50,
      divisor: 100,
    },
  },
  {
    id: "39845",
    name: "Coke",
    price: {
      amount: 70,
      divisor: 100,
    },
  },
  {
    id: "32844",
    name: "Oranges",
    price: {
      amount: 199,
      divisor: 100,
      perKg: true,
    },
  },
  {
    id: "34242",
    name: "Bananas",
    price: {
      amount: 199,
      divisor: 100,
      perKg: true,
    },
  },
];

export default createSlice({
  name: "items",
  initialState: ITEMS,
  reducers: {},
});

export const selectItemsLookup = createSelector(
  (state: RootState) => state.items,
  (items) =>
    items.reduce(
      (map, item) => ({ ...map, [item.id]: item }),
      {} as Record<string, Item>
    )
);

export const selectAllItems = (state: RootState) => state.items;

export const selectItembyId = createSelector(
  selectItemsLookup,
  (_: RootState, { id }: { id: string }) => id,
  (map, id) => map[id]
);
