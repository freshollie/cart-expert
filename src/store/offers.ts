import { createSelector, createSlice } from "@reduxjs/toolkit";
import { create } from "domain";
import { RootState } from ".";
import { Offer } from "../types";

// Offers are in a list as there could be more than 1 discount to apply on an item
const OFFERS: readonly Offer[] = [
  {
    itemId: "34583",
    qualifier: 3,
    name: "3 for 2 on all Beans",
    for: 2,
  },
  {
    itemId: "39845",
    qualifier: 2,
    name: "2 for Â£1 on all coke",
    price: {
      amount: 100,
      divisor: 100,
    },
  },
];

export default createSlice({
  initialState: OFFERS,
  name: "offers",
  reducers: {},
});

export const selectOffers = (state: RootState) => state.offers;
