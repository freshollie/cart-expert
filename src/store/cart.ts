import { createSelector, createSlice, PayloadAction } from "@reduxjs/toolkit";
import { RootState } from ".";
import { selectItemsLookup } from "./items";

const initialState = {} as Record<string, number>;
export default createSlice({
  name: "cart",
  initialState,
  reducers: {
    add: (
      state,
      { payload: { id, amount = 1 } }: PayloadAction<{ id: string; amount?: number }>
    ) => ({
      ...state,
      [id]: (state[id] ?? 0) + amount,
    }),
    remove: (
      state,
      { payload: { id, amount = 1 } }: PayloadAction<{ id: string; amount?: number }>
    ) => ({
      ...state,
      [id]: (state[id] ?? 0) - amount > 0 ? state[id]! - amount : 0,
    }),
    reset: () => initialState,
  },
});

export const selectCartAmount = (itemId: string) => (state: RootState) =>
  state.cart[itemId] ?? 0;

export const selectCartItems = createSelector(
  [(state: RootState) => state.cart, selectItemsLookup],
  (cart, items) =>
    Object.entries(cart)
      .map(([id, amount]) => ({ item: items[id], amount }))
      .filter(({ item, amount }) => item && amount > 0)
      .sort((item1, item2) => item1.item.id.localeCompare(item2.item.id))
);
