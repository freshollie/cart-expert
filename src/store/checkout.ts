import { createSelector } from "@reduxjs/toolkit";
import { isPriceOffer, Offer, Price } from "../types";
import { selectCartItems } from "./cart";
import { selectOffers } from "./offers";

export type Discount = { offer: Offer, savings: Price };
type CheckoutDetails = {
  total: Price;
  subTotal: Price;
  discounts: Discount[];
};

export const selectCheckoutDetails = createSelector(
  [selectCartItems, selectOffers],
  (cart, offers): CheckoutDetails => {
    // find all the discounts which apply to this cart
    const discounts = offers
      .map((offer) => {
        const cartItem = cart.find(({ item }) => offer.itemId === item.id);
        const numDiscounts = Math.floor(
          (cartItem?.amount ?? 0) / offer.qualifier
        );

        if (cartItem && numDiscounts > 0) {
          return new Array(numDiscounts).fill(0).map(() => ({
            offer,
            savings: {
              amount:
                offer.qualifier * cartItem.item.price.amount -
                (isPriceOffer(offer)
                  ? offer.price.amount
                  : offer.for * cartItem.item.price.amount),
              divisor: 100,
            },
          }));
        }
        return [];
      })
      .flat();

    // create totals and subtotals
    const subTotal = {
      amount: cart.reduce(
        (acc, { amount, item }) => acc + amount * item.price.amount,
        0
      ),
      divisor: 100,
    };
    const total = {
      amount:
        subTotal.amount -
        discounts.reduce((acc, discount) => acc + discount.savings.amount, 0),
      divisor: 100,
    };

    return {
      discounts,
      subTotal,
      total,
    };
  }
);
