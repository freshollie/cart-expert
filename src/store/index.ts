import cart from "./cart";
import items from "./items";
import { configureStore } from "@reduxjs/toolkit";
import offers from "./offers";

const store = configureStore({
  reducer: {
    cart: cart.reducer,
    items: items.reducer,
    offers: offers.reducer
  },
});

export type RootState = ReturnType<typeof store.getState>;

export default store;