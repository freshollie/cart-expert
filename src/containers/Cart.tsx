import React, { Fragment } from "react";
import { useSelector } from "react-redux";
import { selectCartItems } from "../store/cart";
import { selectCheckoutDetails } from "../store/checkout";
import { selectItemsLookup } from "../store/items";
import { isPriceOffer, Offer, Price } from "../types";
import "./Cart.css";

const formatOfferPrice = (offer: Offer): string | number => {
  if (isPriceOffer(offer)) {
    const price = offer.price.amount / offer.price.divisor;
    return Math.floor(price) === price ? `Â£${price}` : `${price.toFixed(2)}p`;
  }
  return offer.for;
};

const toPriceValue = (price: Price): number => price.amount / price.divisor;

const formatPrice = (price: Price): string => toPriceValue(price).toFixed(2);

const Cart: React.FC = () => {
  const items = useSelector(selectCartItems);
  const itemLookup = useSelector(selectItemsLookup);
  const { total, subTotal, discounts } = useSelector(selectCheckoutDetails);

  return (
    <div className="cart-container" data-testid="cart">
      <h2>Cart</h2>
      {items.length > 0 ? (
        <table className="cart-details">
          <tbody>
            {items.map(({ item, amount }) =>
              new Array(item.price.perKg ? 1 : amount).fill(0).map((_, i) => (
                <Fragment key={item.id + i}>
                  <tr>
                    <td>{item.name}</td>
                    <td>{!item.price.perKg && formatPrice(item.price)}</td>
                  </tr>
                  {item.price.perKg && (
                    <tr key={item.id + "kg"}>
                      <td>
                        {amount.toFixed(3)} kg @ {toPriceValue(item.price)}/kg
                      </td>
                      <td>
                        {formatPrice({
                          amount: item.price.amount * amount,
                          divisor: item.price.divisor,
                        })}
                      </td>
                    </tr>
                  )}
                </Fragment>
              ))
            )}
            <tr>
              <td>---</td>
              <td></td>
            </tr>
            <tr>
              <td></td>
              <td></td>
            </tr>
            <tr>
              <td>
                <b>Sub Total</b>
              </td>
              <td>
                <b>{formatPrice(subTotal)}</b>
              </td>
            </tr>
            {discounts.map(({ offer, savings }, i) => (
              <tr key={JSON.stringify(offer) + i}>
                <td>
                  {itemLookup[offer.itemId].name} {offer.qualifier} for{" "}
                  {formatOfferPrice(offer)}
                </td>
                <td>-{formatPrice(savings)}</td>
              </tr>
            ))}
            {discounts.length > 0 && (
              <>
                <tr>
                  <td>----</td>
                  <td></td>
                </tr>
                <tr>
                  <td>
                    <b>Total Savings</b>
                  </td>
                  <td>
                    -
                    {formatPrice({
                      amount: subTotal.amount - total.amount,
                      divisor: total.divisor,
                    })}
                  </td>
                </tr>
              </>
            )}
            <tr>
              <td className="break" />
              <td className="break" />
            </tr>
            <tr>
              <td>
                <b>Total to pay</b>
              </td>
              <td>
                <b>{formatPrice(total)}</b>
              </td>
            </tr>
          </tbody>
        </table>
      ) : (
        <b data-testid="cart-no-items-message">No items in cart</b>
      )}
    </div>
  );
};

export default Cart;
