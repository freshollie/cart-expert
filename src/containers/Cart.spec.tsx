import { render } from "@testing-library/react";
import React from "react";
import { Provider } from "react-redux";
import createMockStore from "redux-mock-store";
import { RootState } from "../store";
import Cart from "./Cart";

const exampleItems = [
  {
    id: "34583",
    name: "Beans",
    price: {
      amount: 50,
      divisor: 100,
    },
  },
  {
    id: "39845",
    name: "Coke",
    price: {
      amount: 70,
      divisor: 100,
    },
  },
  {
    id: "32844",
    name: "Oranges",
    price: {
      amount: 199,
      divisor: 100,
      perKg: true,
    },
  },
];

const exampleOffers = [
  {
    itemId: "34583",
    qualifier: 3,
    name: "3 for 2 on all Beans",
    for: 2,
  },
  {
    itemId: "39845",
    qualifier: 2,
    name: "2 for £1 on all coke",
    price: {
      amount: 100,
      divisor: 100,
    },
  },
];

const mockStore = createMockStore<RootState>();

describe("Cart", () => {
  it("should handle displaying the current items in the cart", async () => {
    const [item1, item2] = exampleItems;
    const { asFragment, getByText } = render(
      <Provider
        store={mockStore({
          items: exampleItems,
          offers: exampleOffers,
          cart: {
            [item1.id]: 1,
            [item2.id]: 1,
          },
        })}
      >
        <Cart />
      </Provider>
    );

    expect(getByText(item1.name)).toBeVisible();
    expect(getByText(item2.name)).toBeVisible();

    expect(asFragment()).toMatchSnapshot();
  });

  it("should display discounts of products", async () => {
    const [item1, item2] = exampleItems;
    const { getByText, asFragment } = render(
      <Provider
        store={mockStore({
          items: exampleItems,
          offers: exampleOffers,
          cart: {
            [item1.id]: 3,
            [item2.id]: 2,
          },
        })}
      >
        <Cart />
      </Provider>
    );

    expect(getByText(`${item1.name} 3 for 2`)).toBeVisible();
    expect(getByText(`${item2.name} 2 for £1`)).toBeVisible();
    expect(asFragment()).toMatchSnapshot();
  });

  it("should display the total of all items correctly", () => {
    const [item1, item2] = exampleItems;
    const { getByText } = render(
      <Provider
        store={mockStore({
          items: exampleItems,
          offers: exampleOffers,
          cart: {
            [item1.id]: 3,
            [item2.id]: 2,
          },
        })}
      >
        <Cart />
      </Provider>
    );

    expect(getByText("2.00")).toBeVisible();
  });

  it("should display a message when there are no items in the cart", () => {
    const [item1] = exampleItems;
    const { getByTestId, asFragment } = render(
      <Provider
        store={mockStore({
          items: exampleItems,
          offers: exampleOffers,
          cart: {
            [item1.id]: 0,
          },
        })}
      >
        <Cart />
      </Provider>
    );

    expect(getByTestId("cart-no-items-message")).toBeVisible();
    expect(asFragment()).toMatchSnapshot();
  });
});
