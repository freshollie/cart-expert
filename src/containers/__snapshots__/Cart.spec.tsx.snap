// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart should display a message when there are no items in the cart 1`] = `
<DocumentFragment>
  <div
    class="cart-container"
    data-testid="cart"
  >
    <h2>
      Cart
    </h2>
    <b
      data-testid="cart-no-items-message"
    >
      No items in cart
    </b>
  </div>
</DocumentFragment>
`;

exports[`Cart should display discounts of products 1`] = `
<DocumentFragment>
  <div
    class="cart-container"
    data-testid="cart"
  >
    <h2>
      Cart
    </h2>
    <table
      class="cart-details"
    >
      <tbody>
        <tr>
          <td>
            Beans
          </td>
          <td>
            0.50
          </td>
        </tr>
        <tr>
          <td>
            Beans
          </td>
          <td>
            0.50
          </td>
        </tr>
        <tr>
          <td>
            Beans
          </td>
          <td>
            0.50
          </td>
        </tr>
        <tr>
          <td>
            Coke
          </td>
          <td>
            0.70
          </td>
        </tr>
        <tr>
          <td>
            Coke
          </td>
          <td>
            0.70
          </td>
        </tr>
        <tr>
          <td>
            ---
          </td>
          <td />
        </tr>
        <tr>
          <td />
          <td />
        </tr>
        <tr>
          <td>
            <b>
              Sub Total
            </b>
          </td>
          <td>
            <b>
              2.90
            </b>
          </td>
        </tr>
        <tr>
          <td>
            Beans 3 for 2
          </td>
          <td>
            -0.50
          </td>
        </tr>
        <tr>
          <td>
            Coke 2 for Â£1
          </td>
          <td>
            -0.40
          </td>
        </tr>
        <tr>
          <td>
            ----
          </td>
          <td />
        </tr>
        <tr>
          <td>
            <b>
              Total Savings
            </b>
          </td>
          <td>
            -0.90
          </td>
        </tr>
        <tr>
          <td
            class="break"
          />
          <td
            class="break"
          />
        </tr>
        <tr>
          <td>
            <b>
              Total to pay
            </b>
          </td>
          <td>
            <b>
              2.00
            </b>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;

exports[`Cart should handle displaying the current items in the cart 1`] = `
<DocumentFragment>
  <div
    class="cart-container"
    data-testid="cart"
  >
    <h2>
      Cart
    </h2>
    <table
      class="cart-details"
    >
      <tbody>
        <tr>
          <td>
            Beans
          </td>
          <td>
            0.50
          </td>
        </tr>
        <tr>
          <td>
            Coke
          </td>
          <td>
            0.70
          </td>
        </tr>
        <tr>
          <td>
            ---
          </td>
          <td />
        </tr>
        <tr>
          <td />
          <td />
        </tr>
        <tr>
          <td>
            <b>
              Sub Total
            </b>
          </td>
          <td>
            <b>
              1.20
            </b>
          </td>
        </tr>
        <tr>
          <td
            class="break"
          />
          <td
            class="break"
          />
        </tr>
        <tr>
          <td>
            <b>
              Total to pay
            </b>
          </td>
          <td>
            <b>
              1.20
            </b>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
